cpuminer.go

```diff
@@ -321,79 +320,57 @@ func (m *CPUMiner) generateBlocks(quit chan struct{}) {
 	ticker := time.NewTicker(333 * time.Millisecond)
 	defer ticker.Stop()
 
+	// Subscribe for block template updates.
+	updateChan := m.bg.RequestTemplateUpdate()
+
 out:
 	for {
 		// Quit when the miner is stopped.
 		select {
 		case <-quit:
 			break out
-		default:
-			// Non-blocking select to fall through
-		}
-
-		// If not in connectionless mode, wait until there is a connection to
-		// at least one other peer since there is no way to relay a found
-		// block or receive transactions to work on when there are no
-		// connected peers.
-		if !m.cfg.PermitConnectionlessMining && m.cfg.ConnectedCount() == 0 {
-			time.Sleep(time.Second)
-			continue
-		}
-
-		// No point in searching for a solution before the chain is
-		// synced.  Also, grab the same lock as used for block
-		// submission, since the current block will be changing and
-		// this would otherwise end up building a new block template on
-		// a block that is in the process of becoming stale.
-		m.submitBlockLock.Lock()
-		curHeight := m.g.chain.BestSnapshot().Height
-		if curHeight != 0 && !m.cfg.IsCurrent() {
-			m.submitBlockLock.Unlock()
-			time.Sleep(time.Second)
-			continue
-		}
-
-		// Choose a payment address at random.
-		rand.Seed(time.Now().UnixNano())
-		payToAddr := m.cfg.MiningAddrs[rand.Intn(len(m.cfg.MiningAddrs))]
-
-		// Create a new block template using the available transactions
-		// in the memory pool as a source of transactions to potentially
-		// include in the block.
-		template, err := m.g.NewBlockTemplate(payToAddr)
-		m.submitBlockLock.Unlock()
-		if err != nil {
-			errStr := fmt.Sprintf("Failed to create new block "+
-				"template: %v", err)
-			minrLog.Errorf(errStr)
-			continue
-		}
+		case template := <-updateChan:
+			// Resubscribe for block template updates after receiving an
+			// update.
+			updateChan = m.bg.RequestTemplateUpdate()
+
+			// If not in connectionless mode, wait until there is a connection
+			// to at least one other peer since there is no way to relay a
+			// found block or receive transactions to work on when there are
+			// no connected peers.
+			if !m.cfg.PermitConnectionlessMining && m.cfg.ConnectedCount() == 0 {
+				time.Sleep(time.Second)
```

Comment:
There is no need to sleep here anymore since this code will only run once a new template is generated.